File: 1463.py
import sys
sys.setrecursionlimit(10**9)
from collections import deque

# X가 3으로 나누어 떨어지면, 3으로 나눈다.
# X가 2로 나누어 떨어지면, 2로 나눈다.
# 1을 뺀다.

def convertToOne(num):
  queue = deque()
  queue.append((num,0))
  visited[num] = True

  while queue:
    current, count = queue.popleft()

    if current == 1:
      return count

    if current % 3 == 0 and visited[current//3] == False:
      visited[current//3] = True
      queue.append((current//3, count+1))

    if current % 2 == 0 and visited[current//2] == False:
      visited[current//2] = True
      queue.append((current//2, count+1))

    if visited[current-1] == False:
      queue.append((current-1, count+1))

N = int(input())
visited = [False for _ in range(N+1)]
print(convertToOne(N))

File: 1463.java
import java.io.*;
import java.util.*;

public class Main {
    
    static class Node {
        int value;
        int count;
        
        public Node(int value, int count) {
            this.value = value;
            this.count = count;
        }
    }
    
    static int N;
    static int[] visited;
    static Queue<Node> queue = new LinkedList<>();
    
    public static void main(String args[]) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        
        N = Integer.parseInt(br.readLine());
        visited = new int[N+1];
        visited[N] = 1;
        queue.offer(new Node(N,0));
        
        while(!queue.isEmpty()) {
            Node current = queue.poll();
            if(current.value == 1) {
                bw.write(String.valueOf(current.count));
                break;
            }
            
            if(current.value % 3 == 0 && visited[current.value/3] == 0) {
                queue.offer(new Node(current.value/3, current.count+1));
                visited[current.value/3] = 1;
            }
            
            if(current.value % 2 == 0 && visited[current.value/2] == 0) {
                queue.offer(new Node(current.value/2, current.count+1));
                visited[current.value/2] = 1;
            }
            
            if(visited[current.value-1] == 0) {
                queue.offer(new Node(current.value-1, current.count+1));
                visited[current.value-1] = 1;
            }
        }
        
        bw.flush();
        bw.close();
    }
}
